components/api/src/main/java/com/hotels/styx/api/service/spi/AbstractStyxService.java

```diff
@@ -0,0 +1,107 @@
+import com.hotels.styx.api.HttpResponse;
+import rx.Observable;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.CREATED;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.FAILED;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.RUNNING;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.STARTING;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.STOPPED;
+import static com.hotels.styx.api.service.spi.StyxServiceStatus.STOPPING;
+import static io.netty.handler.codec.http.HttpResponseStatus.OK;
+import static java.lang.String.format;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static rx.Observable.just;
+
+/**
+ * A helper class for implementing StyxService interface.
+ *
+ * AbstractStyxService provides service state management facilities
+ * for implementing a StyxSerive interface.
+ */
+public abstract class AbstractStyxService implements StyxService {
+    private final String name;
+    private final AtomicReference<StyxServiceStatus> status = new AtomicReference<>(CREATED);
+
+    public AbstractStyxService(String name) {
+        this.name = name;
+    }
+
+    public StyxServiceStatus status() {
+        return status.get();
+    }
+
+    protected CompletableFuture<Void> startService() {
+        return completedFuture(null);
+    }
+
+    protected CompletableFuture<Void> stopService() {
+        return completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<Void> start() {
+        boolean changed = status.compareAndSet(CREATED, STARTING);
+
+        if (changed) {
+            return startService()
+                    .exceptionally(failWithMessage("Service failed to start."))
+                    .thenAccept(na -> status.compareAndSet(STARTING, RUNNING));
+        } else {
+            throw new IllegalStateException(format("Start called in %s state", status.get()));
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> stop() {
+        boolean changed = status.compareAndSet(RUNNING, STOPPING);
+
+        if (changed) {
+            return stopService()
+                    .exceptionally(failWithMessage("Service failed to stop."))
+                    .thenAccept(na -> status.compareAndSet(STOPPING, STOPPED));
+        } else {
+            throw new IllegalStateException(format("Stop called in %s state", status.get()));
+        }
+    }
+
+    private Function<Throwable, Void> failWithMessage(String message) {
+        return cause -> {
+            status.set(FAILED);
+            throw new ServiceFailureException(message, cause);
+        };
+    }
+
+    @Override
+    public Map<String, HttpHandler> adminInterfaceHandlers() {
+        return ImmutableMap.of("status", request -> just(HttpResponse.Builder
```

Comment:
This is really meant to be overridden by implemented subclasses. But in case they don't provide meaningful admin interface handlers then the status endpoint alone is generated.

TBH I don't think the handlers are actually populated in the Styx admin interface page yet. This work will be concluded when the SPI interface is opened up.